src/config.rs

```diff
@@ -147,7 +152,51 @@ pub(crate) struct InternalConfig {
     pub(crate) external_port: Option<u16>,
     pub(crate) external_ip: Option<IpAddr>,
     pub(crate) upnp_lease_duration: Duration,
-    pub(crate) min_retry_duration: Duration,
+    pub(crate) retry_config: RetryConfig,
+}
+
+/// Retry configurations for establishing connections and sending messages.
+/// Determines the retry behaviour of requests, by setting the back off strategy used.
+#[derive(Clone, Debug, Copy, Serialize, Deserialize)]
+pub struct RetryConfig {
+    /// The initial retry interval.
+    ///
+    /// This is the first delay before a retry, for establishing connections and sending messages.
+    /// The subsequent delay will be decided by the `retry_delay_multiplier`.
+    pub initial_retry_interval: Duration,
+    /// The maximum value of the back off period. Once the retry interval reaches this
+    /// value it stops increasing.
+    ///
+    /// This is the longest duration we will have,
+    /// for establishing connections and sending messages.
+    /// Retrying continues even after the duration times have reached this duration.
+    /// The number of retries before that happens, will be decided by the `retry_delay_multiplier`.
+    /// The number of retries after that, will be decided by the `retrying_max_elapsed_time`.
+    pub max_retry_interval: Duration,
+    /// The value to multiply the current interval with for each retry attempt.
+    pub retry_delay_multiplier: f64,
+    /// The randomization factor to use for creating a range around the retry interval.
+    ///
+    /// A randomization factor of 0.5 results in a random period ranging between 50% below and 50%
+    /// above the retry interval.
+    pub retry_delay_rand_factor: f64,
+    /// The maximum elapsed time after instantiating
+    ///
+    /// Retrying continues until this time has elapsed.
+    /// The number of retries before that happens, will be decided by the other retry config options.
+    pub retrying_max_elapsed_time: Duration,
```

Comment:
Another reason I ended up with the `min_retry_duration` name is that [the implementation](https://docs.rs/backoff/0.3.0/src/backoff/exponential.rs.html#111-124) uses it like a minimum:

```rust
match self.max_elapsed_time {
    Some(v) if self.get_elapsed_time() > v => None,
    _ => {
        let random = rand::random::<f64>();
        let randomized_interval = Self::get_random_value_from_interval(
            self.randomization_factor,
            random,
            self.current_interval,
        );
        self.current_interval = self.increment_current_interval();
        Some(randomized_interval)
    }
}
```

E.g., retry until more than `max_elapsed_time` has passed, so it's really the minimum amount of time.

Reply:
I think it could be argued that's a bug upstream.

Reply:
That's because it would be very bad to put

`if self.get_elapsed_time() == v`

..since you would then probably never hit that.

So, such cases always use `>=` or `>`, even though it's conceptually a max time that is modeled.

`Max` elapsed time is more accurate because there is not going to be any more time spent retrying than that.
If the retried call succeeds though, it can be a lot shorter than that, so `Min` is not very accurate.

I think they have it right there.

Reply:
I think the implementation is fine, it's just the name sets the wrong expectation for me. "Max elapsed time" makes me think it will not retry for longer than that, when in reality it will retry for _at least_ that long, but in fact strictly longer.